app/queue.py

```diff
@@ -0,0 +1,188 @@
+fake.add_provider(NotifyProvider)
+
+
+def generate_notification():
+    while True:
+        yield fake.notification()
+
+
+def generate_notifications(count=10):
+    notifications = generate_notification()
+    return [next(notifications) for i in range(0, count)]
+
+
+class Buffer(Enum):
+    INBOX = "INBOX"
+    IN_FLIGHT = "IN-FLIGHT"
+
+
+class Queue(ABC):
+    """Queue interface for custom buffer.
+
+    Implementations should allow to poll from the queue and acknowledge
+    read messages once work is done on these.
+    """
+
+    @abstractmethod
+    def poll(self, count=10) -> list[Any]:
+        """Gets messages out of the queue.
+
+        Args:
+            count (int, optional): Number of messages to get out of the queue. Defaults to 10.
+        Returns:
+            list[Any]: List of messages in the queue, from 1 up to {count} number.
+        """
+        pass
+
+    @abstractmethod
+    def acknowledge(self, receipt: UUID):
+        """Acknowledges reception and processing of provided messages IDs.
+
+        Once the acknowledgement is done, the messages will get their in-flight
+        status removed and will not get served again through the `poll` method.
+
+        Args:
+            message_ids (list[int]): [description]
+        """
+        pass
+
+    @abstractmethod
+    def publish(self, notification: Dict) -> None:
+        pass
+
+
+# TODO: Check if we want to move the queue API and implementations into the utils project.
+class RedisQueue(Queue):
+    """Implementation of a queue using Redis."""
+
+    def __init__(self, connection: RedisClient) -> None:
+        self.connection = connection
+        self.limit = current_app.config["BATCH_INSERTION_CHUNK_SIZE"]
+
+    def poll(self, count=10) -> list[Any]:
+        in_flight_key = self.__get_inflight_name()
+        notifications = None
+
+        pipeline = self.connection.pipeline()
+        notifications = pipeline.lrange(Buffer.INBOX, 0, self.limit)
+        pipeline.rpush(in_flight_key, notifications)
+        pipeline.ltrim(Buffer.INBOX, self.limit, -1)
+        pipeline.execute()
+
+        return notifications
+
+    def acknowledge(self, receipt: UUID):
+        inflight_name = self.__get_inflight_name(receipt)
+        self.connection.delete(inflight_name)
+
+    def publish(self, notification: Dict) -> None:
+        self.connection.rpush(Buffer.INBOX, notification)
+
+    def __in_flight(self) -> list[Any]:
+        return self.connection.lrange(Buffer.INBOX, 0, self.limit)
+
+    def __get_inflight_name(self, receipt: UUID = uuid4()) -> str:
+        return f"{Buffer.IN_FLIGHT}:{receipt}"
+
+
+class MockQueue(Queue):
+    """Implementation of a queue that spits out randomly generated notifications.
+
+    Do not use in production!"""
+
+    def poll(self, count=10) -> list[Any]:
+        return generate_notifications(count)
+
+    def acknowledge(self, receipt: UUID):
+        pass
+
+    def publish(self, notification: Dict) -> None:
+        pass
```

Comment:
In the MockQueue it looks like we can publish `Dicts` into it, but we get out `Notification`s?

Reply:
hmm we should change the parameter here for `dict: Dict` ideally. We could pass in anything, as long as it can be a `dict` I think, makes sense? 

Probably we could change the `poll` function to retrieve a `list[dict]` to be consistent too.

Reply:
actually, should we be pushing/pulling the encrypted notification data?

Reply:
We can handle the encryption within the queue itself, on a per config basis if necessary, so it's transparent on your end. We can tackle this in the next iterations.